### PR TITLE
SceneApp: Components for building scene pages in app plugins

### PR DESCRIPTION
**Pull request by dprokop**
Based of https://github.com/grafana/grafana/pull/61746/files#diff-6fa0d2e370daf913de925cfebc01da033392fc381d64012a01fab9b046e7a088

TODO:
- [ ] Docs
- [ ] Tests

### CODE REVIEW

File: src/components/SceneApp/SceneApp.tsx [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+}
+
+export interface SceneAppPageState extends SceneObjectStatePlain {
+  title: string;
+  url: string;
+  routePath?: string;
+  subTitle?: string;
+  hideFromBreadcrumbs?: boolean;
+  tabs?: SceneAppPage[];
+  getScene: (routeMatch: SceneRouteMatch) => EmbeddedScene;
```
Comment (dprokop) [inline src/components/SceneApp/SceneApp.tsx:52]: @torkelo - I've changed this type to return sth more specific than SceneObject - primarily to expose the API for URLSyncManange init.

File: src/components/SceneApp/SceneApp.tsx [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+    const routeMatch = useRouteMatch();
+    console.log('routeMatch path', routeMatch.url);
+
+    let scene = sceneCache.get(routeMatch.url);
+    if (!scene) {
+      scene = model.state.getScene(routeMatch);
+      sceneCache.set(routeMatch.url, scene);
+    }
+
+    scene.initUrlSync();
```
Comment (dprokop) [inline src/components/SceneApp/SceneApp.tsx:109]: Here's the init required prior scene initialisation, ref https://github.com/grafana/scenes/pull/15

File: src/components/SceneApp/SceneApp.tsx [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+    if (drilldowns) {
+      for (const drilldown of drilldowns) {
+        console.log('registering non-tab drilldown route', drilldown.routePath);
+        routes.push(
+          <Route key={drilldown.routePath} exact={false} path={drilldown.routePath}>
+            <SceneAppDrilldownViewRender drilldown={drilldown} parent={model} />
+          </Route>
+        );
+      }
+    }
```
Comment (dprokop) [inline src/components/SceneApp/SceneApp.tsx:98]: Added this for scenarios where drilldown is not part of tabbed page.

File: src/components/SceneApp/SceneApp.tsx (hunk 1 above)
Comment (torkelo) [reply src/components/SceneApp/SceneApp.tsx:52]: makes sense

File: src/components/SceneApp/SceneApp.tsx (hunk 2 above)
Comment (torkelo) [reply src/components/SceneApp/SceneApp.tsx:109]: not sure this is safe / good place, this is in the render path which can happen multiple times, and it causes state changes which are not allowed during react render. 

We might need to move sceneCache.get, model.state.getScene, and scene.initUrlSync to activate on SceneAppPage (and set scene obj in SceneAppState). And only when we have it in state we render it. Might have to clear it from state on deactivate so we don't render it again on re-active before SceneAppPage activate has a change to call scene.initUrlSync.

File: src/components/SceneApp/SceneApp.tsx (hunk 3 above)
Comment (torkelo) [reply src/components/SceneApp/SceneApp.tsx:98]: :+1:

File: src/components/SceneApp/SceneApp.tsx (hunk 2 above)
Comment (dprokop) [reply src/components/SceneApp/SceneApp.tsx:109]: Trickier than expected, will continue next week on this.

File: src/components/SceneApp/SceneApp.tsx [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+   * We use this flag to make sure the URL sync is enabled before the scene is actually rendered.
+   */
+  const [isInitialized, setIsInitialized] = useState(false);
+  const params = useAppQueryParams();
+  const routeMatch = useRouteMatch();
+
+  let scene = sceneCache.get(routeMatch!.url);
+  if (!scene) {
+    // If we are rendering a tab, we need to get the scene f  rom the tab, otherwise, use page's scene
+    scene = activeTab ? activeTab.state.getScene(routeMatch) : page.state.getScene(routeMatch);
```
Comment (dprokop) [inline src/components/SceneApp/SceneApp.tsx:173]: To correctly render active tabs, i.e. when they are part of the drill down.

File: src/components/SceneApp/SceneApp.tsx [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+    // Before rendering scene components, we are making sure the URL sync is enabled for.
+    if (!isInitialized && scene) {
+      scene.initUrlSync();
+      setIsInitialized(true);
+    }
+  }, [isInitialized, scene]);
+
+  if (!isInitialized) {
+    return null;
+  }
```
Comment (dprokop) [inline src/components/SceneApp/SceneApp.tsx:187]: Scene is now read from cache as deep as the scene page is actually rendered. I don't see a need to get it earlier, as the SceneApp and SceneAppPage are primarily responsible for providing routing.

**Review by torkelo — APPROVED** [review]

**Pull request merged**